vms/rpcchainvm/vm_client.go

```diff
@@ -653,6 +654,31 @@ func (vm *VMClient) Disconnected(nodeID ids.ShortID) error {
 	return err
 }
 
+func (vm *VMClient) GetValidatorsByBlockID(blockID ids.ID) (validators.Set, error) {
+	res, err := vm.client.LoadValidators(context.Background(), &vmproto.LoadValidatorsRequest{
+		BlkId: blockID[:],
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not get load validators response: %w", err)
+	}
+	if len(res.ValidatorIds) != len(res.Weights) {
+		return nil, fmt.Errorf("mismatch between validators and weights (%d != %d)", len(res.ValidatorIds), len(res.Weights))
+	}
+	validators := validators.NewSet()
+	for i, validatorId := range res.ValidatorIds {
+		validatorID, err := ids.ToShortID(validatorId)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse validator ID: %w", err)
+		}
+		weight := res.Weights[i]
+		err = validators.AddWeight(validatorID, weight)
+		if err != nil {
+			return nil, fmt.Errorf("could not add validator weight: %w", err)
+		}
```

Comment:
I would have probably changed this method to either be renamed or be split in two, because I don't like that `AddWeight` implicitly creates a validator in the set if we attempt to add a weight to a missing validator, and I especially don't like when it is used specifically for this purpose like it is here.

Upon reading this I initially wondered why a `validators.NewSet` which I presumed empty would already have a validators without their weights at this point. The error string also "pretends" we're only setting a weight here, while we're actually adding a validator to the set.

IMO this should be changed to something like:

```suggestion
		err = validators.Add(validatorID, weight)
		if err != nil {
			return nil, fmt.Errorf("could not add validator: %w", err)
		}
```

And I would make `AddWeight` return an error when attempting to add a weight to a missing validator. This would result in a more consistent and clear API, IMO.

Reply:
I agree entirely, but this is original Avalanche code that I will not touch :)